File: apps/macos/Sources/PaperBananaMac/Services/PaperBananaAPIClient.swift
import Foundation

enum PaperBananaAPIError: LocalizedError {
  case invalidURL(String)
  case invalidResponse
  case server(String)
  case decoding(String)

  var errorDescription: String? {
    switch self {
    case .invalidURL(let value): "无效后端地址：\(value)"
    case .invalidResponse: "后端返回了无效响应。"
    case .server(let message): message
    case .decoding(let message): "响应解析失败：\(message)"
    }
  }
}

final class PaperBananaAPIClient {
  private let session: URLSession
  private let decoder = JSONDecoder()

  init(session: URLSession = .shared) {
    self.session = session
  }

  func fetchBackendHealth(apiBase: String) async throws -> BackendHealth {
    let base = AppDefaults.normalizedAPIBase(apiBase)
    let candidates = [
      try lafEndpoint(apiBase: base),
      try endpoint(apiBase: base, path: "health"),
      try endpoint(apiBase: base, path: "api/health")
    ]

    var lastError: Error?
    for candidate in candidates {
      do {
        let response: HealthEnvelope = try await requestJSON(candidate)
        if let health = response.toBackendHealth(candidateURL: candidate.absoluteString) {
          return health
        }
        throw PaperBananaAPIError.server("当前地址不是 PaperBanana 后端")
      } catch {
        lastError = error
      }
    }
    throw lastError ?? PaperBananaAPIError.server("后端暂时不可用")
  }

  func createJob(apiBase: String, health: BackendHealth?, payload: JobCreatePayload) async throws -> (id: String, status: String) {
    let base = AppDefaults.normalizedAPIBase(apiBase)
    if shouldUsePaperBananaEndpoint(apiBase: base, health: health) {
      let body: [String: Any] = [
        "action": "createJob",
        "configurationMode": payload.configurationMode.rawValue,
        "provider": payload.provider.rawValue,
        "apiKeys": payload.apiKeysBody(),
        "methodContent": payload.methodContent,
        "caption": payload.caption,
        "infographicCategory": payload.infographicCategory,
        "mainModelName": payload.mainModelName,
        "imageModelName": payload.imageModelName,
        "pipelineMode": payload.lafPipelineMode,
        "retrievalSetting": payload.retrievalSetting,
        "aspectRatio": payload.aspectRatio,
        "numCandidates": payload.numCandidates,
        "maxCriticRounds": payload.maxCriticRounds,
        "mock": payload.mock
      ]
      let response: CreateJobEnvelope = try await requestJSON(
        try lafEndpoint(apiBase: base),
        method: "POST",
        body: body
      )
      return (response.jobID ?? response.id ?? "", response.status ?? "queued")
    }

    let body: [String: Any] = [
      "provider": payload.provider.rawValue,
      "configuration_mode": payload.configurationMode.rawValue,
      "api_keys": payload.apiKeysBody(),
      "task_name": "diagram",
      "method_content": payload.methodContent,
      "caption": payload.caption,
      "infographic_category": payload.infographicCategory,
      "main_model_name": payload.mainModelName,
      "image_gen_model_name": payload.imageModelName,
      "pipeline_mode": payload.pipelineMode,
      "retrieval_setting": payload.retrievalSetting,
      "aspect_ratio": payload.aspectRatio,
      "num_candidates": payload.numCandidates,
      "max_critic_rounds": payload.maxCriticRounds,
      "mock": payload.mock
    ]
    let response: CreateJobEnvelope = try await requestJSON(
      try endpoint(apiBase: base, path: "api/jobs"),
      method: "POST",
      body: body
    )
    return (response.id ?? response.jobID ?? "", response.status ?? "queued")
  }

  func getJob(apiBase: String, health: BackendHealth?, jobID: String) async throws -> Job {
    let base = AppDefaults.normalizedAPIBase(apiBase)
    if shouldUsePaperBananaEndpoint(apiBase: base, health: health) {
      let response: JobEnvelope = try await requestJSON(
        try lafEndpoint(apiBase: base),
        method: "POST",
        body: ["action": "getJob", "jobId": jobID]
      )
      if let job = response.job { return job }
      throw PaperBananaAPIError.server("后端没有返回任务详情。")
    }
    return try await requestJSON(try endpoint(apiBase: base, path: "api/jobs/\(jobID)"))
  }

  func userJobs(apiBase: String, health: BackendHealth?) async throws -> [Job] {
    let base = AppDefaults.normalizedAPIBase(apiBase)
    if health?.backendMode == .gateway || shouldUsePaperBananaEndpoint(apiBase: base, health: health) {
      let response: JobsEnvelope = try await requestJSON(
        try lafEndpoint(apiBase: base),
        method: "POST",
        body: ["action": "myJobs", "limit": 50]
      )
      return response.jobs ?? []
    }
    let response: JobsEnvelope = try await requestJSON(try endpoint(
      apiBase: base,
      path: "api/jobs",
      queryItems: [
        URLQueryItem(name: "scope", value: "mine"),
        URLQueryItem(name: "limit", value: "50")
      ]
    ))
    return response.jobs ?? []
  }

  func getSession(apiBase: String) async throws -> CurrentUser? {
    let response: SessionEnvelope? = try? await requestJSON(try endpoint(apiBase: apiBase, path: "api/auth/get-session"))
    return response?.user
  }

  func signIn(apiBase: String, email: String, password: String) async throws {
    try await requestVoid(
      try endpoint(apiBase: apiBase, path: "api/auth/sign-in/email"),
      method: "POST",
      body: ["email": email, "password": password]
    )
  }

  func signUp(apiBase: String, email: String, password: String, name: String) async throws {
    try await requestVoid(
      try endpoint(apiBase: apiBase, path: "api/auth/sign-up/email"),
      method: "POST",
      body: ["email": email, "password": password, "name": name]
    )
  }

  func signOut(apiBase: String) async {
    try? await requestVoid(
      try endpoint(apiBase: apiBase, path: "api/auth/sign-out"),
      method: "POST",
      body: [:]
    )
  }

  private func shouldUsePaperBananaEndpoint(apiBase: String, health: BackendHealth?) -> Bool {
    if health?.backendMode == .fastapi { return false }
    return true
  }

  private func lafEndpoint(apiBase: String) throws -> URL {
    let base = AppDefaults.normalizedAPIBase(apiBase)
    if base.hasSuffix("/paperbanana-api") {
      guard let url = URL(string: base) else { throw PaperBananaAPIError.invalidURL(base) }
      return url
    }
    return try endpoint(apiBase: base, path: "paperbanana-api")
  }

  private func endpoint(apiBase: String, path: String, queryItems: [URLQueryItem]? = nil) throws -> URL {
    let base = AppDefaults.normalizedAPIBase(apiBase)
    guard var components = URLComponents(string: base) else {
      throw PaperBananaAPIError.invalidURL(base)
    }
    let cleanPath = path.hasPrefix("/") ? String(path.dropFirst()) : path
    let existingPath = components.path.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
    let joinedPath = [existingPath, cleanPath]
      .filter { !$0.isEmpty }
      .joined(separator: "/")
    components.path = "/" + joinedPath
    components.queryItems = queryItems
    guard let url = components.url else { throw PaperBananaAPIError.invalidURL(base) }
    return url
  }

  private func requestVoid(_ url: URL, method: String, body: [String: Any]?) async throws {
    let _: EmptyEnvelope = try await requestJSON(url, method: method, body: body)
  }

  private func requestJSON<T: Decodable>(_ url: URL, method: String = "GET", body: [String: Any]? = nil) async throws -> T {
    var request = URLRequest(url: url, timeoutInterval: 30)
    request.httpMethod = method
    request.httpShouldHandleCookies = true
    request.setValue("application/json", forHTTPHeaderField: "Accept")

    if let body {
      request.httpBody = try JSONSerialization.data(withJSONObject: body)
      request.setValue("application/json", forHTTPHeaderField: "Content-Type")
    }

    let (data, response) = try await session.data(for: request)
    guard let httpResponse = response as? HTTPURLResponse else {
      throw PaperBananaAPIError.invalidResponse
    }

    if let serverMessage = serverErrorMessage(from: data), !serverMessage.isEmpty {
      if httpResponse.statusCode < 200 || httpResponse.statusCode >= 300 || serverMessage != "0" {
        throw PaperBananaAPIError.server(serverMessage)
      }
    }

    guard httpResponse.statusCode >= 200 && httpResponse.statusCode < 300 else {
      throw PaperBananaAPIError.server("HTTP \(httpResponse.statusCode)")
    }

    if data.isEmpty {
      return try decoder.decode(T.self, from: Data("{}".utf8))
    }

    do {
      return try decoder.decode(T.self, from: data)
    } catch {
      throw PaperBananaAPIError.decoding(error.localizedDescription)
    }
  }

  private func serverErrorMessage(from data: Data) -> String? {
    guard !data.isEmpty,
          let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
    else { return nil }
    if let code = object["code"] {
      let codeString = String(describing: code)
      if codeString != "0" {
        return object["error"] as? String
          ?? object["detail"] as? String
          ?? object["message"] as? String
          ?? codeString
      }
    }
    if let error = object["error"] as? [String: Any], let message = error["message"] as? String {
      return message
    }
    if let error = object["error"] as? String {
      return error
    }
    return nil
  }
}

private struct EmptyEnvelope: Decodable {}

private struct CreateJobEnvelope: Decodable {
  let id: String?
  let jobID: String?
  let status: String?

  private enum CodingKeys: String, CodingKey {
    case id
    case jobID = "jobId"
    case status
  }
}

private struct JobEnvelope: Decodable {
  let job: Job?
}

private struct JobsEnvelope: Decodable {
  let jobs: [Job]?
}

private struct SessionEnvelope: Decodable {
  let user: CurrentUser?
}

private struct HealthEnvelope: Decodable {
  let ok: Bool?
  let runtime: String?
  let mockEnabled: Bool?
  let laf: NestedHealth?

  private enum CodingKeys: String, CodingKey {
    case ok
    case runtime
    case mockEnabled = "mock_enabled"
    case laf
  }

  func toBackendHealth(candidateURL: String) -> BackendHealth? {
    if runtime == "gateway" {
      return BackendHealth(
        backendMode: .gateway,
        runtime: "gateway",
        mockEnabled: mockEnabled ?? laf?.mockEnabled ?? false,
        detail: laf?.runtime == "laf" ? "Gateway -> Laf" : candidateURL
      )
    }
    if runtime == "laf" {
      return BackendHealth(
        backendMode: .laf,
        runtime: "laf",
        mockEnabled: mockEnabled ?? false,
        detail: candidateURL
      )
    }
    if ok == true {
      return BackendHealth(
        backendMode: .fastapi,
        runtime: runtime ?? "fastapi",
        mockEnabled: mockEnabled ?? false,
        detail: candidateURL
      )
    }
    return nil
  }
}

private struct NestedHealth: Decodable, Equatable {
  let ok: Bool?
  let runtime: String?
  let mockEnabled: Bool?

  private enum CodingKeys: String, CodingKey {
    case ok
    case runtime
    case mockEnabled = "mock_enabled"
  }
}

File: apps/windows/PaperBananaApiClient.cs
using System.Net;
using System.Net.Http.Json;
using System.Runtime.InteropServices.WindowsRuntime;
using System.Text;
using System.Text.Json;
using Windows.Storage.Streams;

namespace PaperBanana.WinUI;

public sealed class PaperBananaApiClient
{
    public const string DefaultApiBase = "https://yifbnnzrwmxn.sealoshzh.site";

    private readonly HttpClient _httpClient;
    private readonly JsonSerializerOptions _jsonOptions = new(JsonSerializerDefaults.Web);

    public PaperBananaApiClient()
    {
        var handler = new HttpClientHandler
        {
            CookieContainer = new CookieContainer(),
            UseCookies = true
        };
        _httpClient = new HttpClient(handler)
        {
            Timeout = TimeSpan.FromSeconds(60)
        };
    }

    public async Task<BackendHealth> FetchBackendHealthAsync(string apiBase, CancellationToken cancellationToken = default)
    {
        var baseUrl = NormalizeApiBase(apiBase);
        var candidates = new[]
        {
            LafEndpoint(baseUrl),
            JoinUrl(baseUrl, "health"),
            JoinUrl(baseUrl, "api/health")
        };

        Exception? lastError = null;
        foreach (var candidate in candidates)
        {
            try
            {
                var root = await SendJsonAsync(candidate, HttpMethod.Get, null, cancellationToken);
                if (TryHealth(root, candidate, out var health)) return health;
                throw new InvalidOperationException("当前地址不是 PaperBanana 后端");
            }
            catch (Exception error)
            {
                lastError = error;
            }
        }

        throw lastError ?? new InvalidOperationException("后端暂时不可用");
    }

    public async Task<(string Id, string Status)> CreateJobAsync(string apiBase, BackendHealth? health, JobCreateRequest payload, CancellationToken cancellationToken = default)
    {
        var baseUrl = NormalizeApiBase(apiBase);
        if (ShouldUsePaperBananaEndpoint(health))
        {
            var body = new Dictionary<string, object?>
            {
                ["action"] = "createJob",
                ["configurationMode"] = payload.ConfigurationMode,
                ["provider"] = payload.Provider,
                ["apiKeys"] = ApiKeysBody(payload.Provider, payload.ApiKey),
                ["methodContent"] = payload.MethodContent,
                ["caption"] = payload.Caption,
                ["infographicCategory"] = payload.InfographicCategory,
                ["mainModelName"] = payload.MainModelName,
                ["imageModelName"] = payload.ImageModelName,
                ["pipelineMode"] = ToLafPipeline(payload.PipelineMode),
                ["retrievalSetting"] = payload.RetrievalSetting,
                ["aspectRatio"] = payload.AspectRatio,
                ["numCandidates"] = payload.NumCandidates,
                ["maxCriticRounds"] = payload.MaxCriticRounds,
                ["mock"] = payload.Mock
            };
            var root = await SendJsonAsync(LafEndpoint(baseUrl), HttpMethod.Post, body, cancellationToken);
            return (ReadString(root, "jobId", "id"), ReadString(root, ["status"], "queued"));
        }

        var fastApiBody = new Dictionary<string, object?>
        {
            ["provider"] = payload.Provider,
            ["configuration_mode"] = payload.ConfigurationMode,
            ["api_keys"] = ApiKeysBody(payload.Provider, payload.ApiKey),
            ["task_name"] = "diagram",
            ["method_content"] = payload.MethodContent,
            ["caption"] = payload.Caption,
            ["infographic_category"] = payload.InfographicCategory,
            ["main_model_name"] = payload.MainModelName,
            ["image_gen_model_name"] = payload.ImageModelName,
            ["pipeline_mode"] = payload.PipelineMode,
            ["retrieval_setting"] = payload.RetrievalSetting,
            ["aspect_ratio"] = payload.AspectRatio,
            ["num_candidates"] = payload.NumCandidates,
            ["max_critic_rounds"] = payload.MaxCriticRounds,
            ["mock"] = payload.Mock
        };
        var fastRoot = await SendJsonAsync(JoinUrl(baseUrl, "api/jobs"), HttpMethod.Post, fastApiBody, cancellationToken);
        return (ReadString(fastRoot, "id", "jobId"), ReadString(fastRoot, ["status"], "queued"));
    }

    public async Task<PaperBananaJob> GetJobAsync(string apiBase, BackendHealth? health, string jobId, CancellationToken cancellationToken = default)
    {
        var baseUrl = NormalizeApiBase(apiBase);
        if (ShouldUsePaperBananaEndpoint(health))
        {
            var root = await SendJsonAsync(
                LafEndpoint(baseUrl),
                HttpMethod.Post,
                new Dictionary<string, object?> { ["action"] = "getJob", ["jobId"] = jobId },
                cancellationToken);
            if (root.TryGetProperty("job", out var job)) return PaperBananaJob.FromJson(job);
            throw new InvalidOperationException("后端没有返回任务详情。");
        }

        return PaperBananaJob.FromJson(await SendJsonAsync(JoinUrl(baseUrl, $"api/jobs/{jobId}"), HttpMethod.Get, null, cancellationToken));
    }

    public async Task<IReadOnlyList<PaperBananaJob>> UserJobsAsync(string apiBase, BackendHealth? health, CancellationToken cancellationToken = default)
    {
        var baseUrl = NormalizeApiBase(apiBase);
        if (ShouldUsePaperBananaEndpoint(health))
        {
            var root = await SendJsonAsync(
                LafEndpoint(baseUrl),
                HttpMethod.Post,
                new Dictionary<string, object?> { ["action"] = "myJobs", ["limit"] = 50 },
                cancellationToken);
            return ReadJobs(root);
        }

        var fastRoot = await SendJsonAsync(JoinUrl(baseUrl, "api/jobs?scope=mine&limit=50"), HttpMethod.Get, null, cancellationToken);
        return ReadJobs(fastRoot);
    }

    public async Task<CurrentUser?> GetSessionAsync(string apiBase, CancellationToken cancellationToken = default)
    {
        try
        {
            var root = await SendJsonAsync(JoinUrl(NormalizeApiBase(apiBase), "api/auth/get-session"), HttpMethod.Get, null, cancellationToken);
            if (!root.TryGetProperty("user", out var user) || user.ValueKind != JsonValueKind.Object) return null;
            var id = ReadString(user, "id");
            if (string.IsNullOrWhiteSpace(id)) return null;
            return new CurrentUser(id, ReadString(user, "email"), ReadString(user, "name"));
        }
        catch
        {
            return null;
        }
    }

    public Task SignInAsync(string apiBase, string email, string password, CancellationToken cancellationToken = default)
    {
        return SendVoidAsync(
            JoinUrl(NormalizeApiBase(apiBase), "api/auth/sign-in/email"),
            new Dictionary<string, object?> { ["email"] = email, ["password"] = password },
            cancellationToken);
    }

    public Task SignUpAsync(string apiBase, string email, string password, string name, CancellationToken cancellationToken = default)
    {
        return SendVoidAsync(
            JoinUrl(NormalizeApiBase(apiBase), "api/auth/sign-up/email"),
            new Dictionary<string, object?> { ["email"] = email, ["password"] = password, ["name"] = name },
            cancellationToken);
    }

    public async Task SignOutAsync(string apiBase, CancellationToken cancellationToken = default)
    {
        try
        {
            await SendVoidAsync(JoinUrl(NormalizeApiBase(apiBase), "api/auth/sign-out"), new Dictionary<string, object?>(), cancellationToken);
        }
        catch
        {
            // Sign-out should clear local UI state even if the server session is already gone.
        }
    }

    public async Task<byte[]> GetImageBytesAsync(string apiBase, string url, CancellationToken cancellationToken = default)
    {
        if (TryReadDataUri(url, out var dataBytes)) return dataBytes;
        var resolved = ResolveImageUrl(apiBase, url);
        return await _httpClient.GetByteArrayAsync(resolved, cancellationToken);
    }

    public string ResolveImageUrl(string apiBase, string url)
    {
        if (string.IsNullOrWhiteSpace(url)) return "";
        if (url.StartsWith("http://", StringComparison.OrdinalIgnoreCase) ||
            url.StartsWith("https://", StringComparison.OrdinalIgnoreCase) ||
            url.StartsWith("data:", StringComparison.OrdinalIgnoreCase) ||
            url.StartsWith("file:", StringComparison.OrdinalIgnoreCase) ||
            url.StartsWith("blob:", StringComparison.OrdinalIgnoreCase))
        {
            return url;
        }
        return $"{NormalizeApiBase(apiBase)}/{url.TrimStart('/')}";
    }

    public static string NormalizeApiBase(string value)
    {
        var trimmed = string.IsNullOrWhiteSpace(value) ? DefaultApiBase : value.Trim();
        return trimmed.TrimEnd('/');
    }

    public static async Task<Microsoft.UI.Xaml.Media.Imaging.BitmapImage> BitmapFromDataUriAsync(string dataUri)
    {
        if (!TryReadDataUri(dataUri, out var bytes)) throw new InvalidOperationException("无效图片数据。");
        var bitmap = new Microsoft.UI.Xaml.Media.Imaging.BitmapImage();
        using var stream = new InMemoryRandomAccessStream();
        await stream.WriteAsync(bytes.AsBuffer());
        stream.Seek(0);
        await bitmap.SetSourceAsync(stream);
        return bitmap;
    }

    private static bool ShouldUsePaperBananaEndpoint(BackendHealth? health)
    {
        return health?.Mode != BackendMode.FastApi;
    }

    private static string LafEndpoint(string apiBase)
    {
        return apiBase.EndsWith("/paperbanana-api", StringComparison.OrdinalIgnoreCase)
            ? apiBase
            : JoinUrl(apiBase, "paperbanana-api");
    }

    private static string JoinUrl(string apiBase, string path)
    {
        if (Uri.TryCreate(path, UriKind.Absolute, out var absolute)) return absolute.ToString();
        return $"{apiBase.TrimEnd('/')}/{path.TrimStart('/')}";
    }

    private async Task SendVoidAsync(string url, object body, CancellationToken cancellationToken)
    {
        await SendJsonAsync(url, HttpMethod.Post, body, cancellationToken);
    }

    private async Task<JsonElement> SendJsonAsync(string url, HttpMethod method, object? body, CancellationToken cancellationToken)
    {
        using var request = new HttpRequestMessage(method, url);
        if (body is not null)
        {
            request.Content = JsonContent.Create(body, options: _jsonOptions);
        }
        return await SendJsonAsync(request, cancellationToken);
    }

    private async Task<JsonElement> SendJsonAsync(HttpRequestMessage request, CancellationToken cancellationToken)
    {
        request.Headers.TryAddWithoutValidation("Accept", "application/json");
        using var response = await _httpClient.SendAsync(request, cancellationToken);
        var text = await response.Content.ReadAsStringAsync(cancellationToken);
        using var document = JsonDocument.Parse(string.IsNullOrWhiteSpace(text) ? "{}" : text);
        var root = document.RootElement.Clone();

        var serverError = ServerError(root);
        if (!response.IsSuccessStatusCode || !string.IsNullOrWhiteSpace(serverError))
        {
            throw new InvalidOperationException(serverError ?? $"HTTP {(int)response.StatusCode}");
        }
        return root;
    }

    private static string? ServerError(JsonElement root)
    {
        if (root.ValueKind != JsonValueKind.Object) return null;
        if (root.TryGetProperty("code", out var code))
        {
            var codeText = code.ValueKind == JsonValueKind.Number ? code.GetInt32().ToString() : code.ToString();
            if (codeText != "0")
            {
                return ReadString(root, "error", "detail", "message", fallback: codeText);
            }
        }
        if (root.TryGetProperty("error", out var error))
        {
            if (error.ValueKind == JsonValueKind.String) return error.GetString();
            if (error.ValueKind == JsonValueKind.Object && error.TryGetProperty("message", out var message)) return message.GetString();
        }
        return null;
    }

    private static bool TryHealth(JsonElement root, string candidateUrl, out BackendHealth health)
    {
        var runtime = ReadString(root, "runtime");
        var mockEnabled = ReadBool(root, "mock_enabled");
        if (runtime == "gateway")
        {
            health = new BackendHealth(BackendMode.Gateway, "gateway", mockEnabled || ReadNestedBool(root, "laf", "mock_enabled"), "Gateway -> Laf");
            return true;
        }
        if (runtime == "laf")
        {
            health = new BackendHealth(BackendMode.Laf, "laf", mockEnabled, candidateUrl);
            return true;
        }
        if (ReadBool(root, "ok"))
        {
            health = new BackendHealth(BackendMode.FastApi, string.IsNullOrWhiteSpace(runtime) ? "fastapi" : runtime, mockEnabled, candidateUrl);
            return true;
        }
        health = new BackendHealth(BackendMode.Gateway, "", false, "");
        return false;
    }

    private static IReadOnlyList<PaperBananaJob> ReadJobs(JsonElement root)
    {
        if (!root.TryGetProperty("jobs", out var jobs) || jobs.ValueKind != JsonValueKind.Array) return [];
        return jobs.EnumerateArray().Select(PaperBananaJob.FromJson).ToArray();
    }

    private static Dictionary<string, string> ApiKeysBody(string provider, string apiKey)
    {
        return new Dictionary<string, string>
        {
            ["openrouter"] = provider == "openrouter" ? apiKey : "",
            ["gemini"] = provider == "gemini" ? apiKey : "",
            ["openai"] = provider == "openai" ? apiKey : "",
            ["bailian"] = provider == "bailian" ? apiKey : ""
        };
    }

    private static string ToLafPipeline(string mode)
    {
        return mode switch
        {
            "demo_full" => "full",
            "vanilla" => "vanilla",
            _ => "planner_critic"
        };
    }

    private static bool TryReadDataUri(string value, out byte[] bytes)
    {
        bytes = [];
        var marker = ";base64,";
        var index = value.IndexOf(marker, StringComparison.OrdinalIgnoreCase);
        if (!value.StartsWith("data:", StringComparison.OrdinalIgnoreCase) || index < 0) return false;
        bytes = Convert.FromBase64String(value[(index + marker.Length)..]);
        return true;
    }

    private static string ReadString(JsonElement element, params string[] names)
    {
        return ReadString(element, names, fallback: "");
    }

    private static string ReadString(JsonElement element, string[] names, string fallback)
    {
        foreach (var name in names)
        {
            if (element.ValueKind == JsonValueKind.Object && element.TryGetProperty(name, out var value))
            {
                return value.ValueKind switch
                {
                    JsonValueKind.String => value.GetString() ?? fallback,
                    JsonValueKind.Number => value.ToString(),
                    JsonValueKind.True => "true",
                    JsonValueKind.False => "false",
                    _ => fallback
                };
            }
        }
        return fallback;
    }

    private static string ReadString(JsonElement element, string first, string second, string fallback = "")
    {
        return ReadString(element, [first, second], fallback);
    }

    private static string ReadString(JsonElement element, string first, string second, string third, string fallback = "")
    {
        return ReadString(element, [first, second, third], fallback);
    }

    private static bool ReadBool(JsonElement element, string name)
    {
        if (!element.TryGetProperty(name, out var value)) return false;
        return value.ValueKind switch
        {
            JsonValueKind.True => true,
            JsonValueKind.False => false,
            JsonValueKind.String => bool.TryParse(value.GetString(), out var parsed) && parsed,
            _ => false
        };
    }

    private static bool ReadNestedBool(JsonElement element, string property, string nestedProperty)
    {
        return element.TryGetProperty(property, out var nested) && nested.ValueKind == JsonValueKind.Object && ReadBool(nested, nestedProperty);
    }
}
